server/gotenberg/gotenberg.go

```diff
@@ -0,0 +1,197 @@
+// Package gotenberg is a Gotenberg client with stream (io.Reader) support.
+package gotenberg
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+)
+
+const (
+	// pingTimeout set while sending ping requests to Gotenberg server.
+	pingTimeout = time.Second * 5
+
+	// defaultConvertTimeout set while sending and waiting response ofr file convert requests
+	// to Gotenberg server.
+	defaultConvertTimeout = time.Minute * 10
+)
+
+const (
+	// pingEndpoint used to status check Gotenberg to see if it's running and ready.
+	pingEndpoint = "/ping"
+
+	// convertEndpoint used to convert files to PDFs.
+	convertEndpoint = "/convert/office"
+)
+
+const multipartField = "file"
+
+// supportedFormats are the supported file formats  that can be converted to PDF by Gotenberg.
+var supportedFormats = []string{"doc", "docx", "odt", "xls", "xlsx", "ods", "ppt", "pptx", "odp"}
+
+// Gotenberg is a client for Gotenberg server.
+// for more info see: https://github.com/thecodingmachine/gotenberg
+// warning! don't forget to set proper timeouts as you need in Gotenberg server because default ones are too low. ex:
+// - docker run -d -p 4798:3000 --env DEFAULT_WAIT_TIMEOUT=600 --env MAXIMUM_WAIT_TIMEOUT=600 thecodingmachine/gotenberg:6
+type Gotenberg struct {
+	m sync.RWMutex // protects config.
+	// config holds configs set as options.
+	conf *config
+}
+
+// config holds configs set as options.
+type config struct {
+	// addr is network address of Gotenberg server.
+	addr string
+	// convertTimout is used during sending file convert requests.
+	convertTimeout time.Duration
+}
+
+// New creates new Gotenberg client with given Gotenberg server addr and options.
+func New(addr string, options ...Option) *Gotenberg {
+	g := &Gotenberg{conf: &config{addr: addr}}
+	g.UpdateConfig(options...)
+	return g
+}
+
+// UpdateConfig updates options set during the first initialization of New.
+func (g *Gotenberg) UpdateConfig(options ...Option) {
```

Comment:
Being able to change the configuration seems to come with the complexity of having to safeguard the config from concurrent access. What if we just forced the client to create a new `Gotenberg` instance in that case instead?

Reply:
Right. We have a concurrency safe implementation anyway but I agree that we can just get rid of `UpdateConfig` and reinitialize the app on every configuration change.

It'll make things even simpler. Also, this is still will be efficient, since configuration updates shouldn't happen very often.

P.S.: I wanted to experiment a new thing that different than what we have in other plugins when I come up with _UpdateConfig_. But it's even better to take reinitialization approach.

Reply:
removed by 3e03bd479501bd15845b7b0793bd601acffda8b2